x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/parser/LogicalPlanBuilder.java

```diff
@@ -221,18 +225,33 @@ public PlanFactory visitStatsCommand(EsqlBaseParser.StatsCommandContext ctx) {
         }
         // grouping keys are automatically added as aggregations however the user is not allowed to specify them
         if (groupings.isEmpty() == false && aggregates.isEmpty() == false) {
-            var groupNames = new LinkedHashSet<>(Expressions.names(Expressions.references(groupings)));
+            var groupNames = new LinkedHashSet<>(Expressions.names(groupings));
+            var groupRefNames = new LinkedHashSet<>(Expressions.names(Expressions.references(groupings)));
 
             for (NamedExpression aggregate : aggregates) {
-                if (Alias.unwrap(aggregate) instanceof UnresolvedAttribute ua && groupNames.contains(ua.name())) {
-                    throw new ParsingException(ua.source(), "Cannot specify grouping expression [{}] as an aggregate", ua.name());
+                Expression e = Alias.unwrap(aggregate);
+                if (e.resolved() == false && e instanceof UnresolvedFunction == false) {
+                    String name = e.sourceText();
+                    if (groupNames.contains(name)) {
+                        fail(e, "Remove grouping key [{}] as is already included in the STATS output", name);
+                    } else if (groupRefNames.contains(name)) {
+                        fail(e, "Cannot specify grouping expression [{}] as an aggregate", name);
+                    }
                 }
             }
         }
-        aggregates.addAll(groupings);
+        // since groupings are aliased, add refs to it in the aggregates
+        for (Expression group : groupings) {
+            aggregates.add(Expressions.attribute(group));
+        }
+
         return input -> new Aggregate(source(ctx), input, new ArrayList<>(groupings), aggregates);
     }
 
+    private void fail(Expression exp, String message, Object... args) {
+        throw new VerificationException(Collections.singletonList(Failure.fail(exp, message, args)));
+    }
```

Comment:
Improved the detection of grouping (or references) added to the aggregate. In case of failure a verification exception is thrown instead of a ParsingException for consistency reasons.